src/Core/src/Stream/RewindableStream.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace AsyncAws\Core\Stream;
+
+use AsyncAws\Core\Exception\InvalidArgument;
+
+/**
+ * Provides a Stream that can be read several time.
+ *
+ * This is for internal use only. One cannot iterate only over a few items of the stream.
+ * If iterating over a stream, the full stream must be consumed before calling methods:
+ * - stringify
+ * - length
+ * - hash
+ *
+ * @internal
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+final class RewindableStream implements Stream
+{
+    private $content;
+
+    /**
+     * @var Stream
+     */
+    private $fallback;
+
+    private function __construct(Stream $content)
+    {
+        $this->content = $content;
+    }
+
+    public static function create($content): RewindableStream
+    {
+        if ($content instanceof self) {
+            return $content;
+        }
+        if ($content instanceof Stream) {
+            return new self($content);
+        }
+
+        throw new InvalidArgument(sprintf('Expect content to be a "Stream". "%s" given.', \is_object($content) ? \get_class($content) : \gettype($content)));
+    }
+
+    public function length(): ?int
+    {
+        if (null !== $this->fallback) {
+            return $this->fallback->length();
+        }
+
+        return $this->content->length();
+    }
+
+    public function stringify(): string
+    {
+        if (null !== $this->fallback) {
+            return $this->fallback->stringify();
+        }
+
+        return \implode('', \iterator_to_array($this));
+    }
+
+    public function getIterator(): \Traversable
+    {
+        if (null !== $this->fallback) {
+            yield from $this->fallback;
+
+            return;
+        }
+
+        $resource = \fopen('php://memory', 'r+b');
+        $this->fallback = ResourceStream::create($resource);
+
+        $size = 0;
+        $inmemory = true;
+        foreach ($this->content as $chunk) {
+            \fwrite($resource, $chunk);
+            if ($inmemory) {
+                $size += \strlen($chunk);
+                // switch to filesystem if string larger that 1 Mb
+                if ($size > 1024 * 1024) {
```

Comment:
You could use `php://temp`, which starts in memory and automatically turns it into a file when reaching a size threshold (2MB by default, but that can be configured when creating the stream). This would simplify the code a lot, as you are basically reimplementing it on top of `php://memory` and `tmpfile()`

Reply:
I wasn't ware of that! thanks for the tips!

Reply:
Cool. I didnt know that either.